### ISSUE TITLE
[DocDB][Tablet Splitting] Scan returns incomplete results with yb_enable_expression_pushdown=on for table with splitted tablets

### ISSUE THREAD

**Issue by arybochkin** [yugabyte/yugabyte-db#13688]
Jira Link: [DB-3233](https://yugabyte.atlassian.net/browse/DB-3233)
### Description

The second child tablet is completely ignored during read operation over a splitted parent tablet when `yb_enable_expression_pushdown=on`: after meta cache invalidation the read operation retry happens only over the first child tablet -- the one whose `partition_start_key` matches parent's `partition_start_key`. The issue is related to #12631, but the fix for #12631 will not cover this case as tablet may be still not splitted when the read requests are composed and sent at PGGate side, and tablet may begin splitting around the time the batcher handles these request at YBClient side.

Steps to reproduce:
```
1. $ ./bin/yb-ctl start --master_flags="enable_automatic_tablet_splitting=false"

2. # create table t1(k1 text not null, v1 text, v2 int, primary key ((k1) hash)) split into 4 tablets;
3. # insert into t1 select i::text, i::text, i from  generate_series(1,70000) as i;

4. $ ./build/latest/bin/yb-admin -master_addresses localhost:7100 list_tablets ysql.yugabyte t1

Tablet-UUID                      	Range                                                    	
62d22015d8084146908e62a99cff4024 	partition_key_start: "" partition_key_end: "?\377"	
f7d3abcb9d724b9e92eebe5e1f06457d 	partition_key_start: "?\377" partition_key_end: "\177\376" 	
0d57389b727443b9904793688088c93f 	partition_key_start: "\177\376" partition_key_end: "\277\375" 	
b89bc7a4f615428da6000bbb43fad561 	partition_key_start: "\277\375" partition_key_end: "" 	

5. # set yb_enable_expression_pushdown=off;
6. # select * from t1 where k1='49199' and v1='49199';
  k1   |  v1   |  v2   
-------+-------+-------
 49199 | 49199 | 49199
(1 row)

7. # set yb_enable_expression_pushdown=on;
8. # select * from t1 where k1='49199' and v1='49199';
  k1   |  v1   |  v2   
-------+-------+-------
 49199 | 49199 | 49199
(1 row)

9. $ ./build/latest/bin/yb-ts-cli flush_tablet 0d57389b727443b9904793688088c93f
10. $ ./build/latest/bin/yb-admin split_tablet -master_addresses localhost:7100 0d57389b727443b9904793688088c93f

11. # set yb_enable_expression_pushdown=off;
12. # select * from t1 where k1='49199' and v1='49199';
  k1   |  v1   |  v2   
-------+-------+-------
 49199 | 49199 | 49199
(1 row)

13. # set yb_enable_expression_pushdown=on;
14. # select * from t1 where k1='49199' and v1='49199';
 k1 | v1 | v2 
----+----+----
(0 rows)

15. $ ./bin/yb-ctl stop && ./bin/yb-ctl start --master_flags="enable_automatic_tablet_splitting=fals

16. # set yb_enable_expression_pushdown=on;
17. # select * from t1 where k1='49199' and v1='49199';
  k1   |  v1   |  v2   
-------+-------+-------
 49199 | 49199 | 49199
(1 row)

```
Note: the value for k1 and v1 must be from the second child after split, its hash code should be in ["\240\206", "\277\375"); set_dump can be used to search for such value.

Findings from the test:
1. Due to "pushdown" specific, the read operation is executed over all tablets of the table
2. The table was created with 4 partitions (== tablets)
3. Due to optimizations, read operations are grouped into two batches (two read op in each batch), each read op has partition_key equal to the corresponding partition_key_start of a tablet
4. Required row was located in the third tablet, then it's "moved" to the second child after the split
5. Neither PGGate nor YBClient are aware of changed partitions during steps 11-14.
6. The error is happening during validating the state of the third tablet as the tablet is in bad state (deleted | being splitted | splitting completed), table is marked as 'stale' and batcher reties the same read operation
7. The metacache is invalidated and refreshed during next lookup and the lookup finds a valid tablet (the first child) by the original partition_key, the problem is that new tablet covers only the half of the original interval defined in the read request (hash_code / max_hash_code)
8. Read operation successfully completed, but second child is completely ignored, and this child contains the required record. The request returns incomplete result.

The questions are:
1. Does returning of paging state in the response for the operation help?
2. If paging state is sent from DocDB, why PGGate does not handle it properly?

Related issues: #6719, #12648, #12631

**Comment by arybochkin** [comment]
The fix for #13660 mask the issue.

**Comment by andrei-mart** [comment]
I believe #13684 and #13660 solve the problem. Would you mind to retest?

**Comment by arybochkin** [comment]
@andrei-mart none of issues is reproduced starting from the fix of #13684.

**Comment by arybochkin** [comment]
This issue is fixed by #13684: the request is now uses `partition_column_values` source instead of `lower `/ `upper` bounds used previously; using of the partition column values allows to specify the exact partition key that should be used for the request forwarding; so, after ybclient detects its hash is stale (due to tablet splitting) and updates the cache, the lookup returns the expected second tablet found by correct partition key (instead of bounds).
As for the original issue: lower bound was used to identify the tablet, and after the meta-cache had been updated the request was sent to the first child tablet, but the tablet was not returning paging state to cover the second child tablet due to the condition `|| !pgsql_read_request.partition_column_values().empty()` inside `Tablet::IsQueryOnlyForTablet()` -- this means if partition columns values are specified then it is supposed that the request is intended for the exactly this tablet; so, probably the tablet is supposed to not return paging state as it shouldn't hold any info about fresh splitting and meta-cache update info.
Several tests have been executed to cover the similar cases, indexes and range partitioning cases, everything looks as expected. So, with the fix of #13684 it's not possible to simulate the previous situation and the fix looks sufficient. Hence closing the issue.
